scalafix-nsc/src/main/scala/scalafix/nsc/NscSemanticApi.scala

```diff
@@ -11,110 +11,466 @@ import scalafix.ScalafixConfig
 import scalafix.rewrite.SemanticApi
 import scalafix.util.logger
 
-case class SemanticContext(enclosingPackage: String, inScope: List[String])
-
 trait NscSemanticApi extends ReflectToolkit {
 
-  /** Returns a map from byte offset to type name at that offset. */
-  private def offsetToType(gtree: g.Tree,
-                           dialect: Dialect): mutable.Map[Int, m.Type] = {
-    // TODO(olafur) Come up with more principled approach, this is hacky as hell.
-    // Operating on strings is definitely the wrong way to approach this
-    // problem. Ideally, this implementation uses the tree/symbol api. However,
-    // while I'm just trying to get something simple working I feel a bit more
-    // productive with this hack.
-    val builder = mutable.Map.empty[Int, m.Type]
-    def add(gtree: g.Tree, ctx: SemanticContext): Unit = {
-
-      /** Removes redudant Foo.this.ActualType prefix from a type */
-      val stripRedundantThis: m.Type => m.Type = _.transform {
-        case m.Term.Select(m.Term.This(m.Name.Indeterminate(_)), qual) =>
-          qual
-        case m.Type.Select(m.Term.This(m.Name.Indeterminate(_)), qual) =>
-          qual
-      }.asInstanceOf[m.Type]
-
-      val stripImportedPrefix: m.Type => m.Type = _.transform {
-        case prefix @ m.Type.Select(_, name)
-            if ctx.inScope.contains(prefix.syntax) =>
-          name
-      }.asInstanceOf[m.Type]
-
-      val stripEnclosingPackage: m.Type => m.Type = _.transform {
-        case typ: m.Type.Ref =>
-          import scala.meta._
-          typ.syntax.stripPrefix(ctx.enclosingPackage).parse[m.Type].get
-      }.asInstanceOf[m.Type]
-
-      val cleanUp: (Type) => Type =
-        stripRedundantThis andThen
-          stripImportedPrefix andThen
-          stripEnclosingPackage
-
-      val parsed = dialect(gtree.toString()).parse[m.Type]
-      parsed match {
-        case m.Parsed.Success(ast) =>
-          builder(gtree.pos.point) = cleanUp(ast)
-        case _ =>
-      }
+  /** The compiler sets different symbols for `PackageDef`s
+    * and term names pointing to that package. Get the
+    * underlying symbol of `moduleClass` for them to be equal. */
+  @inline
+  private def getUnderlyingPkgSymbol(pkgSym: g.Symbol) = {
+    if (!pkgSym.isModule) pkgSym
+    else pkgSym.asModule.moduleClass
+  }
+
+  /** Keep members that are relevant for scoping rules.
+    * For example, remove methods (don't appear in types).  */
+  @inline
+  private def keepRelevantMembers(s: g.Symbol): Boolean =
+    (s.isModule || s.isValue || s.isAccessor || s.isType) &&
+      !(s.isMethod || s.isSynthetic || s.hasPackageFlag)
+
+  @inline
+  private def mixScopes(sc: g.Scope, sc2: g.Scope) = {
+    val mixedScope = sc.cloneScope
+    sc2.foreach(s => mixedScope.enterIfNew(s))
+    mixedScope
+  }
```

Comment:
We should optimize for correctness and simplicity for now. We can worry about performance later.